### PR TITLE
Initial implementation

### PR DESCRIPTION
**Pull request by mike183**
Initial implementation for the "stand-alone" engine blueprint.

cc/ @rwjblue @trentmwillis

### CODE REVIEW

**Comment by mike183** [comment]
This should be ready for a review.

**Review by rwjblue — APPROVED** [review]

File: files/routable-files/index.js [hunk 1]
```diff
@@ -0,0 +1,11 @@
+/*jshint node:true*/
```
Comment (stefanpenner) [inline files/routable-files/index.js]: `'use strict';`

File: files/routable-files/index.js [hunk 2]
```diff
@@ -0,0 +1,11 @@
...
+var EngineAddon = require('ember-engines/lib/engine-addon');
+module.exports = EngineAddon.extend({
+  name: '<%= dasherizedModuleName %>',<% if (hasLazyFlag) { %>
+
+  lazyLoading: <%= isLazy %>,<% } %>
+
+  isDevelopingAddon: function() {
```
Comment (stefanpenner) [inline files/routable-files/index.js]: `isDevelopingAddon() { }`

and this should likely only be true if `in-repo`

File: files/routeless-files/index.js [hunk 3]
```diff
@@ -0,0 +1,11 @@
+/*jshint node:true*/
```
Comment (stefanpenner) [inline files/routeless-files/index.js]: `'use strict';`

File: files/routeless-files/index.js [hunk 4]
```diff
@@ -0,0 +1,11 @@
...
+var EngineAddon = require('ember-engines/lib/engine-addon');
```
Comment (stefanpenner) [inline files/routeless-files/index.js]: `const`

File: files/routeless-files/index.js [hunk 5]
```diff
@@ -0,0 +1,11 @@
...
+module.exports = EngineAddon.extend({
+  name: '<%= dasherizedModuleName %>',<% if (hasLazyFlag) { %>
+
+  lazyLoading: <%= isLazy %>,<% } %>
+
+  isDevelopingAddon: function() {
```
Comment (stefanpenner) [inline files/routeless-files/index.js]: same as above re: `isDevelopingAddon`

File: index.js [hunk 6]
```diff
@@ -0,0 +1,147 @@
...
+    if (existsSync(engineFilePath)) {
+      return engineFilePath;
+    } else if (existsSync(addonFilePath)) {
+      return addonFilePath;
+    } else {
+      return path.resolve(this._appBlueprint.path, 'files', file);
+    }
+  },
+
+  generatePackageJson: function(options, isInstall) {
```
Comment (stefanpenner) [inline index.js]: concise function

**Review by stefanpenner — CHANGES_REQUESTED** [review]

File: files/routable-files/index.js (hunk 1 above)
Comment (mike183) [reply files/routable-files/index.js]: Added

File: files/routable-files/index.js (hunk 2 above)
Comment (mike183) [reply files/routable-files/index.js]: Removed the `isDevelopingAddon` method.

File: files/routeless-files/index.js (hunk 3 above)
Comment (mike183) [reply files/routeless-files/index.js]: Added

File: files/routeless-files/index.js (hunk 4 above)
Comment (mike183) [reply files/routeless-files/index.js]: Updated, and also in the `routable-files/index.js` file.

File: files/routeless-files/index.js (hunk 5 above)
Comment (mike183) [reply files/routeless-files/index.js]: Removed the `isDevelopingAddon` method.

File: index.js (hunk 6 above)
Comment (mike183) [reply index.js]: Updated

**Comment by mike183** [comment]
Thanks for the review @stefanpenner, I believe I've addressed all of the issues you mentioned.

File: blueprints/engine-blueprint/index.js [hunk 7]
```diff
@@ -0,0 +1,23 @@
+'use strict';
+const path = require('path');
+const MainBlueprint = require('../../../index');
+
+/*
+  Create an "addon blueprint" that simply defers to our
```
Comment (trentmwillis) [inline blueprints/engine-blueprint/index.js]: Could we add a `TODO` marker to this comment block? Just so it is obvious that there is something incomplete with this.

File: blueprints/engine-blueprint/index.js [hunk 8]
```diff
@@ -0,0 +1,23 @@
...
+    * ember-addon key in package.json
+    * ember-addon-main.js file
+    * blueprints/ folder
+ */
+module.exports = Object.assign({}, MainBlueprint, {
+  init() {
+    this._super.init.apply(this, arguments);
+
+    this.path = path.join(__dirname, '..', '..', '..');
+    this.name = 'ember-engines/engine-blueprint';
```
Comment (trentmwillis) [inline blueprints/engine-blueprint/index.js:21]: Potentially dumb question, can we not set these properties in the `MainBlueprint` and avoid the `init` altogether?

File: files/routable-files/index.js [hunk 9]
```diff
@@ -0,0 +1,9 @@
+/*jshint node:true*/
+'use strict';
+
+const EngineAddon = require('ember-engines/lib/engine-addon');
+
+module.exports = EngineAddon.extend({
+  name: '<%= dasherizedModuleName %>'<% if (hasLazyFlag) { %>,
+  lazyLoading: <%= isLazy %><% } %>
```
Comment (trentmwillis) [inline files/routable-files/index.js]: Should this be updated to use the new object definition for `lazyLoading`? Same for the routeless engine blueprint as well.

File: files/routable-files/addon-config/environment.js [hunk 10]
```diff
@@ -0,0 +1,11 @@
+/*jshint node:true*/
+'use strict';
+
+module.exports = function(environment) {
+  var ENV = {
```
Comment (trentmwillis) [inline files/routable-files/addon-config/environment.js]: Can use `const` here I believe.

File: files/routeless-files/addon-config/environment.js [hunk 11]
```diff
@@ -0,0 +1,11 @@
+/*jshint node:true*/
+'use strict';
+
+module.exports = function(environment) {
+  var ENV = {
```
Comment (trentmwillis) [inline files/routeless-files/addon-config/environment.js]: `const` here as well.

**Review by trentmwillis — CHANGES_REQUESTED** [review]
Overall looks good to me, just a bunch of minor things (some for my own education).

File: blueprints/engine-blueprint/index.js (hunk 7 above)
Comment (mike183) [reply blueprints/engine-blueprint/index.js]: Added

File: blueprints/engine-blueprint/index.js (hunk 8 above)
Comment (mike183) [reply blueprints/engine-blueprint/index.js:21]: I think this is all just work-around code that will be removed when the mentioned issue is resolved and this file is removed.

Though, I just copied this from the glimmer blueprint so I could be wrong. I guess @rwjblue will be able to answer if this should be moved or not.

File: files/routable-files/index.js (hunk 9 above)
Comment (mike183) [reply files/routable-files/index.js]: Updated. I was actually thinking about this just as I finished up for the day yesterday.

@trentmwillis / @rwjblue: Should we also update this in the `in-repo-engine` blueprint as well?

File: files/routable-files/addon-config/environment.js (hunk 10 above)
Comment (mike183) [reply files/routable-files/addon-config/environment.js]: Updated

File: files/routeless-files/addon-config/environment.js (hunk 11 above)
Comment (mike183) [reply files/routeless-files/addon-config/environment.js]: Updated

**Comment by mike183** [comment]
Thanks @trentmwillis, I've addressed all of the issues you mentioned and also left a question for you regarding the `in-repo-engine` blueprint.

File: files/routable-files/index.js (hunk 9 above)
Comment (rwjblue) [reply files/routable-files/index.js]: Yes, I think we should.

Comment (mike183) [reply files/routable-files/index.js]: PR'd: https://github.com/ember-engines/ember-engines/pull/430

**Review by trentmwillis — APPROVED** [review]
Looks good to me. The above comment doesn't need to block this from landing as it could be addressed when the other `TODO` is addressed as well.

**Pull request merged**